### DISCUSSION TITLE
Going semi-open source

### DISCUSSION THREAD

**Discussion by Wertik** [Zrips/CMI#4568] [Q&A]
**Question:**

Due to how many open issues there are and how much this plugin covers, I think it would be a great step forward to provide source to at least a closed circle of developers to help speed up fixing issues and implementing more new features.

So my question is, do you agree and will you do it?

**Comment by CloudeLecaw** [comment]
u may shoudl wrote Zrips diectly in discord and not creat a ticket for this ;)

**Comment by Wertik** [comment]
> u may shoudl wrote Zrips diectly in discord and not creat a ticket for this ;)

Well it's a general suggestion and here it can gather opinions from others. Also I doubt Zrips would ever see my DM. His discord has to be spammed hard.

**Comment by Wertik** [comment]
A good implementation of this would be creating a bounty system similar to what Songoda does. But I bet people will fix issues and/or implement new features even for free.

**Comment by MoonSoD** [comment]
Well, looking at the current number of issues (990) the suggestion seems pretty on point. Possibly a big leap forward for CMI.

**Comment by Zrips** [comment]
Going semi-open source on a paid plugin is a strange way to do things. I mean, yea, there are people who could fix things better then i do or help out in fixing most minor things, as those can take as much time in testing as bigger ones while fix itself could be quite simple. Thing would be in actually properly setting things up for some random people to have access to actual source code. Even tho CMI is not obfuscated, so its not like i'm hiding its code or something... But well, again, circling back to the main point, its a paid plugin and having someone to help out on its development for free is some what strange. 
Tho, it is interesting idea.

**Reply by tinyoverflow** [reply]
While I absolutely do get your point, your plugin is on sale for a one-time payment of 15 bucks. It's not that much.
I for myself, for example, am a person who doesn't want to reinvent the wheel. So I grab plugins that are either open source or where I can get the source after the purchase so I can bring that into my own GitHub and continue developing it so it suits my needs. When possible, I merge the upstream into my own copy to both benefit from my contributions as well as from the contributions of the original developer or the community.

This also ensures that I can continue to use the plugin when the original maintainer decides to let that plugin die.

I'm very interested in this plugin, but the fact that you do not release the source either to the public or at least to customers makes this decission really hard for me. Yes, I could decompile and work on it. But then you release a new version. I would have to decompile it again and then what? Bringing that to git is really hard as I cannot simply pull your changes and rebase my work on it.

**Reply by Wertik** [reply]
You're taking this a bit too far. I'm all for open source work, after all, that's what this platform has been built for. But making your work, that you want to earn money for, accessible to everyone is something I wouldn't have done, the amount of contributors would probably never outgrow the people who just compile to get it for free or repost. Hence -- semi-open source, a term that I used out of nowhere and probably doesn't exist. 

What I wanted to encourage is including more people internally to work on the plugin, because I saw the state and the number of open issues and feature requests. Zrips is the developer of many more plugins that are vial for survival servers, having more people would increase the quality and time it takes to fix issues and implement new features.

**Reply by tinyoverflow** [reply]
I think I get what you're saying, but I would like to clarify some things that I think some people my not realize:
Open Source *does not* mean "It is free, do what you want". Of course open source projects can be licensed and there are some great examples of paid open source software:
- Nextcloud is Open Source, but the important features are paid.
- Proxmox is also Open Source. It is btw. a german company. They make their money by selling support packages.
- Mattermost is an Open Source chat server. While I believe that it offers a free plan, the most of its features are paid.

You obviously have to keep in mind that these solutions mostly offer a free plan, but in general they're not made for home use, but for enterprise use. But the concept stays the same.

Redistributing can be prohibited through licensing. Any redistribution without permission can be taken down and he could theoretically sue him for a damage compensation. Just compiling it by yourself and using it without a license can be done, but would be forbidden to do so. GriefDefender, for example, explicitly allows that to test it before purchase. If you're happy, purchase it.

Of course, there are people who don't follow this. But I think these are mostly people who cannot afford that or would download that from sites like BS anyway.

**Reply by mrfloris** [reply]
Just to be 'that guy' .. Just because others have done it, doesn't mean everybody should though.. 

That said, I am not against open source, but it's currently a premium plugin. CMILib going open source and CMI not being obfuscated, and there being a CMI API, I think that would complete the 'semi open source' part of things. 

I rather know that one or two people from a company work on something, than depend on that person to vet all the code that others suggest and hope no backdoors or exploits are being sneaked in.. Not to mention having to support all those forks of forks who have weird custom builds (and probably sell them... or at least release them) .. and then them coming to ask and expecting support. 

For me personally: how it is right now, fine by me.. It might help with some support tickets or bug reports, but it doesn't mean that it will resolve the majority of them. Looking at other spigotmc projects that are well known, open sourced, etc. Their support isn't magically resolved by people jumping on projets. Generally all I see if a few helping for their personal problem and then we never heard from them again. And as soon as someone who's done some significant contributions doesn't agree with the author, they disappear and split and start their own branch and splitting things up ~ counterproductive to trying to work together. 

If however ever in the future zrips decides to call it quits or has lost interest in further developing the plugin, and therefore wouldn't mind having a drop in income from it either .. releasing it open source for others to keep it current .. seems like a welcomed plan-b to  help prevent it from dying.

**Comment by 456dev** [comment]
There are some projects that have successfully implemented this, by only providing compiled jarsfiles to those who pay, because most people would pay for the convenience of having a jarsfile premade

**Reply by mrfloris** [reply]
Quite a few of those do not make the same numbers as I have seen CMI (talking spigot stats). One is giving up quite a bit of guaranteed income as well probably.

**Comment by Wertik** [comment]
Yeah, you're talking about a good point, I've been thinking about this too. But that's also why I suggested going only "semi-". What you could do is pick out a closed group of developers throughout the community, who are insterested and are skilled enough to help you maintain the project on a bigger scale. (only talking about 5-10, maybe even less people)

About people working for free on a paid plugin, this is your choice and theirs. You sure make some money from your plugins (or just CMI). A better model for this would probably be creating bounties and letting people choose to do it for money -- or do it for free.

Setting this up could be as simple as creating a private discord server with a join room (for requests and getting a role), then on github either setting up an organization (better for access maintanence than single contributors).

Now, more points I'd like to make is that you could use this to learn a bit more about how to work in a team. It also opens up a way to meet more developers. There have been a lot of teams and orgs forming up recently, if you meet the right people, you could take your plugins even further.

**Comment by CloudeLecaw** [comment]
Sometimes 5 people is leaking too much. There are enough free plugins that can do the same. And if someone dont want to pay this little fee for cmi. He has the choose to take the other plugins ;)
I think zrips has a little group allready that he Trust in.

**Reply by tinyoverflow** [reply]
I think you're getting a little confused here. Open Source does not mean free. Open Source means collaboration.

**Reply by Wertik** [reply]
And as I stated in previous posts, this would rather mean having a **closed** group of people.

**Reply by mrfloris** [reply]
A lot of people hear open source and instantly think "free to download jar file, since i can compile it myself" <- and they're not wrong in that aspect.

**Comment by CloudeLecaw** [comment]
well i hope cmi does not become open source xD

**Reply by tinyoverflow** [reply]
For what reasons?

**Reply by mrfloris** [reply]
I have no problem with CMI becoming open source at some point, if not for portions of it.

**Comment by CloudeLecaw** [comment]
> Open source is source code that is made freely available for possible modification and redistribution. Products include permission to use the source code,design documents, or content of the product.

CMI should not open its source so others can build their own nonsense on its code or write free modifications of it. CMI is good as it is ;;-)

Why not simply develop your own source? then you can give release it and others can use it *g*
What a luck that premium plugins are not forced to release their source code to others. this would hurt the whole implementation of spigots premium sources.
Today u have a premium - just wait a few days and someone releases a modificated free version based on its source code,.... no thanks @.@

**Reply by tinyoverflow** [reply]
There is something called licensing. You can also reply to single comments and do not have to reply to the whole topic, btw.

**Reply by 456dev** [reply]
> so others can build their own nonsense on its code

Cmi already has an API [here](https://github.com/zrips/cmi-api) 
How is it different?

**Comment by CloudeLecaw** [comment]
maybe but noone needs to release it source code anymore today if not wanted.. what a luck

Well may u see .. im not a friend of the idea of open source xD

May Bill branded me *hehe*

**Comment by TomLewis** [comment]
A lot of large pluigns have their code open source but only offer a jar for those that pay. 

Thai works great as it's quite hard to compile a plugin unless you have done it before and have setup everything.

It wil only be posative allowing people to fix things and add commits.

Your workload is massive and there's just too many suggestions and issue's for you to keep up with! 


It does open the door to new customers, people that do compile it for free and use it, piracy actually helps sell movies/TV shows.

I'm a paying customer by the way! I'm just an advocate for open source!

**Comment by mrfloris** [comment]
CMILib could be a first step towards being more open source, and it would make more sense as well since Jobs and other open source projects depend on it - it feels weird having open source projects with closed libs behind it. 

But as a business decision to keep things closed source, perfectly fine by me. It's easy to decompile, and reported bugs are handled at some point.

**Comment by IceColdTray** [comment]
Freemium is also an option. Plugin is still paid on resource websites like SpigotMC but you can compile from source but support is only for buyers.

**Reply by gecko10000** [reply]
As Floris said though, a significant amount of purchases would be lost.

**Reply by ghost** [reply]
Most people would rather just fork over some money to receive a compiled jar than to jump through the hurdles of getting gradle set up. You almost certainly lose more profits from BlackSpigot/piracy now than open sourcing would ever cost you. Additionally, most people will buy the plugin if it means guaranteed automatic updates, regular bug fixes, and support. 

Most importantly, no plugin is perfect. Allowing people to peer review your code can be essential for finding vulnerabilities, incompatibilities, and poorly optimized code.

**Reply by gecko10000** [reply]
That's true; I know I would contribute, haha

**Reply by akarisama** [reply]
> As Floris said though, a significant amount of purchases would be lost.

I disagree with this personally. The concept of lost purchases is kind of a fallacy. If someone is willing to pirate software, they're not the type who would ever pay for anything, no matter what. A pirate will use either pirate CMI, pirate another essentials type plugin, or go with something free.

Similarly, there's a lot of people that shy away from purchasing closed source plugins because of the large amount of abandoned plugins on spigot. If you're a developer (and believe me, no one else knows what gradle even is, let alone how to use it), having available source could give you the confidence you need to pick CMI over alternatives. Not only that, but compiling code every time you want to update CMI is a lot of hoops to go through even for a developer, and they're hoops you'll have to continuously go through unless you buy the full product.

**Reply by mrfloris** [reply]
Sorry, in a way I can see what you mean, but the question is not "go open source or it will be pirated" .. The reasons for open source are not about fighting piracy. Piracy will happen no matter what.

**Comment by TomLewis** [comment]
The biggest benefit would be people creating features that we've been asking for years that Zrips doesn't have time to do, the plugin has so much potential and not enough time in the day!

Maybe a better option would be to allow a few good developers to help out do features (unpaid) as contributors that want to expand the plugin for the community, then its not entirely open source but still has extra sets of hands (that are vetted first to be good programmers).

I wouldn't mind seeing someone dedicated to keeping the website updated too, sometimes we miss features due to lack of documentation, if someone has the time to read through all the code and update documentation for missing features that would also be beneficial to everyone.

**Reply by mrfloris** [reply]
But if there's no time, then people who PR things etc, should also realise that for their code to merge there has to be time to review and test and adjust, optimize etc as well. And even then, looking at something like Residence or Jobs, I hardly see any real features being added through PR, or significant contributions (minus a few exceptions). 

That said, I understand what you mean as well, i dont disagree with some of the things you say.

**Reply by TomLewis** [reply]
Then have a testing team.

I have an entire team of peolpe to test, and a dedicated test server.

**Reply by mrfloris** [reply]
Life is not that black and white.
